Suppress compilation warnings CS1701, CS1702, and CS1705

### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched [issues](https://github.com/cake-build/cake/issues) to ensure it has not already been reported

### Cake runner

Cake .NET Tool, Cake Frosting, Cake runner for .NET Framework, Cake runner for .NET Core

### Cake version

1.1.0

### Operating system

N/A

### Operating system architecture

N/A

### CI Server

N/A

### What are you seeing?

When Cake addins reference assemblies that have already been loaded by Cake, and the version loaded in the assembly load context doesn't match the version targeted by the Cake addin, the user sees a warning written to the Console (even with the default Verbosity).

For example, using the latest version (as of this writing) of the [Cake.ExtendedNuGet](https://github.com/cake-contrib/Cake.ExtendedNuGet) addin (v4.0.2) with the latest version of Cake (v1.1.0) produces the following warning:

> (3145,12): warning CS1701: Assuming assembly reference 'NuGet.Versioning, Version=5.4.0.3, Culture=neutral, PublicKeyToken=31bf3856ad364e35' used by 'Cake.ExtendedNuGet' matches identity 'NuGet.Versioning, Version=5.8.1.7021, Culture=neutral, PublicKeyToken=31bf3856ad364e35' of 'NuGet.Versioning', you may need to supply runtime policy

This happens because Cake references a newer version of the NuGet packages (v5.8.1) whilst the latest version of the `Cake.ExtendedNuGet` addin references an older version of NuGet (v5.4.0) which is still compatible, but different. More details described in https://github.com/cake-contrib/Cake.ExtendedNuGet/issues/35#issuecomment-833761870

### What is expected?

These warnings should be suppressed given that they provide little value to the Cake user (i.e. NuGet upgrades are a common scenario when loading transitive dependencies) and there's nothing that the user can do about it other than ignoring the "noise".

There seems to be general agreement in the .NET community and within Microsoft that these warnings should be suppressed by default and potentially completely removed (1).

These warnings have had to be disabled in project templates for a long time. Searching for "NoWarn" 1701 on GitHub at the moment yields over 500K results: https://github.com/search?q=NoWarn+1701&type=Code&utf8=%E2%9C%93

I propose suppressing the warnings [`CS1701`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs1701), [`CS1702`](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1702) and [`CS1705`](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs1705) completely, or at the very least, changing its severity so that it only appears when running Cake with `Diagnostic` verbosity.

- (1)
  - [[Scripting] Suppress CS1701 by default](https://github.com/dotnet/roslyn/issues/5501)
  - [Warnings CS1701 and CS1702 should be removed](https://github.com/dotnet/roslyn/issues/19640)
  - [Missing default NoWarn settings when NoWarn is defined in Directory.Build.props](https://github.com/dotnet/sdk/issues/1205#issuecomment-302579464)

### Steps to Reproduce

1. Add the following to a `build.cake` file:

```
#addin nuget:?package=Cake.ExtendedNuGet&version=4.0.2
```

2. Run `dotnet cake`


### Output log

_No response_

--- a/src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs
+++ b/src/Cake/Infrastructure/Scripting/RoslynScriptSession.cs
@@ -115,6 +115,30 @@ namespace Cake.Infrastructure.Scripting
 
             foreach (var diagnostic in diagnostics)
             {
+                // Suppress some diagnostic information. See https://github.com/cake-build/cake/issues/3337
+                switch (diagnostic.Id)
+                {
+                    // CS1701 Compiler Warning (level 2)
+                    // Assuming assembly reference "Assembly Name #1" matches "Assembly Name #2", you may need to supply runtime policy
+                    // https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-messages/cs1701
+                    case "CS1701":
+                        continue;
+
+                    // CS1702 Compiler Warning (level 3)
+                    // Assuming assembly reference "Assembly Name #1" matches "Assembly Name #2", you may need to supply runtime policy
+                    // https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs1702
+                    case "CS1702":
+                        continue;
+
+                    // CS1705 Compiler Error
+                    // Assembly 'AssemblyName1' uses 'TypeName' which has a higher version than referenced assembly 'AssemblyName2'
+                    case "CS1705":
+                        continue;
+
+                    default:
+                        break;
+                }
+
                 switch (diagnostic.Severity)
                 {
                     case DiagnosticSeverity.Info:

